src/graph.rs

```diff
@@ -0,0 +1,924 @@
+
+            let last_user_count = match self.users_per_second.last() {
+                Some(last) => *last,
+                None => 0,
+            };
+            expand_per_second_metric_array(&mut self.users_per_second, second, last_user_count);
+            self.users_per_second[second] = users;
+        }
+    }
+
+    /// Generate active users graph.
+    pub(crate) fn get_active_users_graph(&self) -> Graph<usize> {
+        Graph::new(
+            "graph-active-users",
+            "Active users #",
+            self.add_timestamp_to_html_graph_data(&self.users_per_second),
+            self.starting,
+            self.started,
+            self.stopping,
+            self.stopped,
+        )
+    }
+
+    /// Generate requests per second graph.
+    pub(crate) fn get_requests_per_second_graph(&self) -> Graph<u32> {
+        Graph::new(
+            "graph-rps",
+            "Requests #",
+            self.add_timestamp_to_html_graph_data(&self.requests_per_second),
+            self.starting,
+            self.started,
+            self.stopping,
+            self.stopped,
+        )
+    }
+
+    /// Generate average response time graph.
+    pub(crate) fn get_average_response_time_graph(&self) -> Graph<u32> {
+        let response_times = self
+            .average_response_time_per_second
+            .iter()
+            .map(|moving_average| moving_average.average as u32)
+            .collect::<Vec<_>>();
+
+        Graph::new(
+            "graph-avg-response-time",
+            "Response time [ms]",
+            self.add_timestamp_to_html_graph_data(&response_times),
+            self.starting,
+            self.started,
+            self.stopping,
+            self.stopped,
+        )
+    }
+
+    /// Generate active tasks graph.
+    pub(crate) fn get_tasks_per_second_graph(&self) -> Graph<usize> {
+        Graph::new(
+            "graph-tps",
+            "Tasks #",
+            self.add_timestamp_to_html_graph_data(&self.tasks_per_second),
+            self.starting,
+            self.started,
+            self.stopping,
+            self.stopped,
+        )
+    }
+
+    /// Generate errors per second graph.
+    pub(crate) fn get_errors_per_second_graph(&self) -> Graph<u32> {
+        Graph::new(
+            "graph-eps",
+            "Errors #",
+            self.add_timestamp_to_html_graph_data(&self.errors_per_second),
+            self.starting,
+            self.started,
+            self.stopping,
+            self.stopped,
+        )
+    }
+
+    /// Adds timestamps to the graph data series to ensure correct time display on x axis.
+    ///
+    /// Will take a vector of (generally numerical) values and convert them into tuples where
+    /// the second element will be the data point and the first element will be formatted time
+    /// it belongs to.
+    fn add_timestamp_to_html_graph_data<T: Copy>(&self, data: &[T]) -> Vec<(String, T)> {
+        // TODO properly handle the date Option timestamps.
```

Comment:
I'm not understanding what's not handled properly here? What's the implication?

Reply:
Yeah, me neither. I was thinking that I should handle the case where those dates are not set yet. But then I realized that this function is never called if that is the case.